File: solutions/Java/MaxSubArray_DP.java
public class MaxSubArray_DP {
    /**
     * Dynamic Programming approach - Build up subarray sums.
     *
     * Time Complexity: O(n)
     * Space Complexity: O(1) - Optimized space version
     *
     * @param nums Array of integers
     * @return Maximum subarray sum
     */
    public static int maxSubArray(int[] nums) {
        int n = nums.length;
        if (n == 1) return nums[0];

        int maxSum = nums[0];
        int dp = nums[0];  // max sum ending at current position

        for (int i = 1; i < n; i++) {
            dp = Math.max(nums[i], dp + nums[i]);
            maxSum = Math.max(maxSum, dp);
        }

        return maxSum;
    }

    public static void main(String[] args) {
        int[] test1 = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
        System.out.println("Test 1: " + maxSubArray(test1)); // Expected: 6
    }
}

File: solutions/C++/max_subarray_dp.cpp
#include <vector>
#include <algorithm>

using namespace std;

int maxSubArray(vector<int>& nums) {
    /**
     * Dynamic Programming approach - Build up subarray sums.
     *
     * Time Complexity: O(n)
     * Space Complexity: O(1) - Optimized space version
     *
     * @param nums Vector of integers
     * @return Maximum subarray sum
     */
    int n = nums.size();
    if (n == 1) return nums[0];

    int max_sum = nums[0];
    int dp = nums[0];  // max sum ending at current position

    for (int i = 1; i < n; i++) {
        dp = max(nums[i], dp + nums[i]);
        max_sum = max(max_sum, dp);
    }

    return max_sum;
}
